### PR TITLE
add ordered-imports rule 

### PR DESCRIPTION
**Pull request by falkenhawk**
with following predefined dynamic grouping config:

1. built-in node modules
2. dependencies (read from project's package.json)
3. aliased paths, which begin with tilde (~) in our convention
4. the rest, incl. typescript absolute imports
5. parent directories
6. current directory

### CODE REVIEW

File: tslint.ts [hunk 1]
```diff
@@ -1,10 +1,74 @@
...
 const rulesAirbnb = tslintAirbnb.rules;
 const rulesReact = tslintReact.rules;
 
+// prepare regex for node core modules for 'ordered-imports' rule
+const modules = Object.entries(coreModules as Record<string, boolean>).reduce(
+  (acc: string[], [module, enabled]) => {
+    enabled && !module.startsWith('_') && acc.push(module);
+    return acc;
+  },
+  []
```
Comment (prk3) [inline tslint.ts:17]: That's a misuse of reduce.
```
.filter(([module, enabled]) => enabled && !module.startsWith('_'))
.map(([module]) => module);
```

Comment (prk3) [reply tslint.ts:17]: Why are we filtering out modules starting with `_` btw?

File: tslint.ts [hunk 2]
```diff
@@ -1,10 +1,74 @@
...
+);
+const coreModulesRegex = '^(' + modules.join('|') + ')$';
```
Comment (prk3) [inline tslint.ts:19]: Are module names escaped at this point?

File: tslint.ts [hunk 3]
```diff
@@ -1,10 +1,74 @@
...
+    ...Object.keys(packageJson.dependencies || {}),
+    ...Object.keys(packageJson.devDependencies || {}),
+  ];
+
+  if (dependencies.length) {
+    dependenciesRegex = '^(' + dependencies.join('|').replace(/\./g, '\\\\.') + ')(/|$)';
+  }
+}
+
+const importGroups = [
```
Comment (prk3) [inline tslint.ts:36]: Conditional concatenation is a bit ugly. You can do:
```
},
condition && {
  some: stuff,
},
```
and then `.filter(Boolean)` on an entire array. It's a lot cleaner.

File: tslint.ts [hunk 4]
```diff
@@ -1,10 +1,74 @@
...
+  (acc: string[], [module, enabled]) => {
+    enabled && !module.startsWith('_') && acc.push(module);
+    return acc;
+  },
+  []
+);
+const coreModulesRegex = '^(' + modules.join('|') + ')$';
+
+// prepare import groups for 'ordered-imports' rule
+const projectPackageJson = path.join(process.cwd(), 'package.json');
```
Comment (visuallization) [inline tslint.ts:22]: @falkenhawk nice! do you know how this affects compile time?

Comment (falkenhawk) [reply tslint.ts:22]: i havent profiled this tbh 🙈 but this shouldn't affect compile/startup time of the app, only `yarn lint` command... or "real-time" checks in IDE

**Review by prk3 — CHANGES_REQUESTED** [review]
Finally :tada:

File: tslint.ts (hunk 1 above)
Comment (Southgarden116) [reply tslint.ts:17]: @prk3 why is it a misuse of reduce? reduce > filter/map always!

Comment (flipace) [reply tslint.ts:17]: IMO this is not a misuse - it prevents going through the list of modules twice, which for something that may run continuously in the background is not a bad thing

File: tslint.ts [hunk 5]
```diff
@@ -1,10 +1,74 @@
...
+}
+
+const importGroups = [
+  {
+    name: 'aliased paths, which begin with tilde in our convention',
+    match: '^~',
+    order: 30,
+  },
+  {
+    name: 'parent directories',
```
Comment (flipace) [inline tslint.ts:43]: will the "farthest" imports be put on top?

File: tslint.ts (hunk 3 above)
Comment (flipace) [inline tslint.ts:36]: i think maybe one thing we could add would be to put imports from scope `@ovos-media/` below third party, but thats just a thought

Comment (prk3) [reply tslint.ts:36]: :+1:

File: tslint.ts (hunk 1 above)
Comment (prk3) [reply tslint.ts:17]: Reduce (fold) is for reducing a list into a single value. Yes, that single value can be an array, but at this point we're basically using is as a `for..of` loop. The cost of iterating through a list of 100 elements twice is negligible (probably less than the cost of expanding array with .push - that's why Array has a constructor taking capacity btw).

File: tslint.ts (hunk 2 above)
Comment (falkenhawk) [reply tslint.ts:19]: they are not... but there aren't any atm which e.g. contain a dot

File: tslint.ts (hunk 3 above)
Comment (falkenhawk) [reply tslint.ts:36]: thanks for the idea!

File: tslint.ts (hunk 5 above)
Comment (falkenhawk) [reply tslint.ts:43]: the import groups are matched first, then ordered by `order` property

File: tslint.ts (hunk 1 above)
Comment (falkenhawk) [reply tslint.ts:17]: > 
> 
> Why are we filtering out modules starting with `_` btw?

there are modules like `_stream_duplex|_stream_transform|_stream_wrap|_stream_passthrough|_stream_readable|_stream_writable` which are not meant to be imported directly

File: tslint.ts (hunk 5 above)
Comment (falkenhawk) [reply tslint.ts:43]: ah you meant the farthest as in the number of `../` - the imports will be ordered alphabetically, so most probably the farthest will be on the top within the group

File: tslint.ts (hunk 1 above)
Comment (falkenhawk) [reply tslint.ts:17]: @prk3 @flipace would a `for..of` loop make more sense here?

Comment (falkenhawk) [reply tslint.ts:17]: I tried `for..of` but I have no idea why typescript is yelling at me :/
image

**Pull request merged**